### ISSUE TITLE
implementation of non-german instances

### ISSUE THREAD

**Issue by Entkenntnis** [serlo/frontend#226]
Adding support for other language instances will need unexpectedly more things to be considered than I anticipated, including:

- missing i18n know-how on my part
- some complexities of the frontend (serverless deployment) and the legacy system (stretched across php and database)
- unfamiliarity with other language versions of serlo

Therefore I suggest to postpone the implementation, as we have already a huge volumen of issues for the german instance and there is no pressure to implement other language instances right now.

This will close #188 and #225.

**Comment by inyono** [comment]
We shouldn’t underestimate the effort to add i18n support to an app that wasn’t build with i18n in mind. It is going to get only more complicated for every feature we add without i18n support. So I wouldn’t push this too far into the future.

**Comment by Entkenntnis** [comment]
That's very true. So it'll be worthwhile to start with some architectural thoughts ...

Whats bugging me a bit at the moment is that the current system has many different levels of translations:

1. php gettext and the .po files (ui strings)
2. The different landing pages (de, en, default) with hardcoded strings
3. Translations from the database (menus, content)
4. This file: [src/config/autload/global.php](https://github.com/serlo/serlo.org/blob/ac35240abfeb3277ee9803fb28dda2fedd851190/packages/public/server/src/config/autoload/global.php)
5. edtr-io with its own system

This makes it very hard for me to decide where and how to translate the frontend. This is part of the complexity that I haven't anticipated ... but I would be very happy to help overhauling and streamlining this ...

**Comment by inyono** [comment]
TL;DR: just ignore the legacy system since you don't work with that (you have the API for this)

> php gettext and the .po files (ui strings)

You can ignore that completely since the frontend replaces the old Ui.

> The different landing pages (de, en, default) with hardcoded strings

Again, just do whatever you currently do in the redesign. In the legacy system, we had basically a default fallback landing page that could be customized if you wanted something special. Might need the same behavior in the redesign in the long-term, but shouldn't affect i18n much.

> Translations from the database (menus, content)

Well, that's user-generated content. Not sure how this affects what you are currently doing in the redesign? No need to translate user-generated content since it is already in the correct language.

> This file: src/config/autload/global.php

Again, ignore it and handle it however you like in the redesign.

> edtr-io with its own system

Isn't Edtr.io specific. That's just the i18n for the client. This is pretty much the only interesting thing for you since you'll probably handle i18n in a similar way. If the custom handling to integrate i18next into Next.js annoys you too much (understandably), you can use something differently. E.g. https://github.com/vinissimus/next-translate looks promising. See also [i18n for Serverless Next.js app](https://github.com/zeit/next.js/discussions/10651)

**Comment by Entkenntnis** [comment]
Thank you for the comments on the legacy system. The meaning of "ignore it" in this context basically means "rebuild it from scratch"? I was hoping to take over at least some part of the existing architecture ... (and not having to rebuild everything)

This issue now comes down to designing a new i18n system for serlo.org. This is more challenging, but that's OK. I will start to make some plans and investigate the issue.

**Comment by inyono** [comment]
> Thank you for the comments on the legacy system. The meaning of "ignore it" in this context basically means "rebuild it from scratch"? I was hoping to take over at least some part of the existing architecture ... (and not having to rebuild everything)

Pretty much, yeah. There's not much architecture you can take over since the current architecture (except i18next) is tied to PHP. Furthermore, there is not much architecture to take over, anyway, since most of the integration is just declaring strings as translatable (and this is completely dependent on which strings you use in your UI).

> This issue now comes down to designing a new i18n system for serlo.org. This is more challenging, but that's OK. I can start to make some plans and investigate the issue (and check how high this is prioritized within Serlo)

Not sure what you mean by designing. We need to integrate a JavaScript i18n library (requirement: automatically extract strings from source files as something that Crowdin understands, most likely JSON) and replace every hard-coded string (that should be translatable) in the frontend code base. Regarding prioritization: This is definitely a blocker for 100% since we have active non-German instances. And like I said above, delaying it makes it more annoying in the future ;)

**Comment by Entkenntnis** [comment]
this is now in the roadmap after #289, #290 and #263

**Comment by Entkenntnis** [comment]
Ok, the i18n implementation is on its way (finally), so this issue can be closed, current step is #365
